### PR TITLE
Honor Retry-able JD Responses

### PR DESCRIPTION
**Pull request by bellmyer**
# "Honor Thy Headers"

John Deere API response codes 429 (Too Many Requests) and 503 (Service Unavailable) mean something has gone wrong, but the request may be retried. These responses may or may not include a Retry-After header requesting a certain wait time.

This client has been updated to do the following for CRUD requests (GET, POST, PUT, DELETE):

* if no Retry-After header is included, increase wait time exponentially (doubling each attempt) from 1 second to ~34 minutes before the next retry.
* if a Retry-After header is included, use the greater of this header value or the exponential wait time for this request count.
* retry 429/503 requests a maximum of 12 times, then return a JD::MaxRetriesExceededError error, which includes the request method and last response code from JD.

The testing for this has been very thorough, but I feel it needed to be. I even had to break down and add Mocha for setting expectations.

# Developer Checklist

- [x] Github label added, milestone added, review requested
- [x] Rebased with latest merge target
- [x] I have commented my code, particularly in hard-to-understand areas, and at the top of new interactors
- [x] I have added tests
- [ ] The build is green

### CODE REVIEW

File: lib/my_john_deere_api/accessor.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+module MyJohnDeereApi
+  class Accessor
+    attr_reader :access_token
+
+    REQUEST_METHODS = [:get, :post, :put, :delete]
+
+    RETRY_DELAY_EXPONENT = 2
+    MAX_RETRIES = 12
```
Comment (allknowingfrog) [inline lib/my_john_deere_api/accessor.rb:8]: Is this configurable? When we make API requests in a job, we probably don't want to wait the full 34 minutes, especially when Sidekiq already has an exponential retry mechanism.

Comment (bellmyer) [reply lib/my_john_deere_api/accessor.rb:8]: I agree, I was thinking that all hardcoded parts should be defaults which can be overridden with an options hash. I'm going to do that.

File: lib/my_john_deere_api/client.rb [hunk 2]
```diff
@@ -45,7 +45,7 @@ def initialize(api_key, api_secret, options = {})
 
     def accessor
       return @accessor if defined?(@accessor)
-      @accessor = OAuth::AccessToken.new(consumer.user_get, access_token, access_secret)
+      @accessor = NetHttpRetry::Decorator.new(OAuth::AccessToken.new(consumer.user_get, access_token, access_secret))
```
Comment (allknowingfrog) [inline lib/my_john_deere_api/client.rb:48]: It's possible that I'm missing something, but it seems like you still need a mechanism to get configuration options from the outside application through the client and down into the retry decorator.

Comment (bellmyer) [reply lib/my_john_deere_api/client.rb:48]: You are correct, sir. I'll make the fix.

**Pull request merged**